File: data_mcp/gpt-5/scala/rep3/src/Main.scala
//> using scala "3.3.1"
//> using dep "org.http4s::http4s-ember-server:0.23.25"
//> using dep "org.http4s::http4s-dsl:0.23.25"
//> using dep "org.http4s::http4s-circe:0.23.25"
//> using dep "io.circe::circe-core:0.14.7"
//> using dep "io.circe::circe-generic:0.14.7"
//> using dep "io.circe::circe-parser:0.14.7"
//> using dep "org.typelevel::cats-effect:3.5.4"

import cats.effect._
import cats.syntax.all._
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.circe._
import org.http4s.circe.CirceEntityCodec._
import io.circe._
import io.circe.syntax._
import io.circe.generic.semiauto._
import java.util.UUID
import org.typelevel.ci.{CIString, CIStringSyntax}
import com.comcast.ip4s._

object Main extends IOApp {

  case class UserRec(id: Int, username: String, password: String)
  private case class UserPublic(id: Int, username: String)
  private implicit val userPublicEncoder: Encoder[UserPublic] = deriveEncoder

  case class TodoRec(
      id: Int,
      ownerId: Int,
      title: String,
      description: String,
      completed: Boolean,
      created_at: String,
      updated_at: String
  )
  private case class TodoPublic(
      id: Int,
      title: String,
      description: String,
      completed: Boolean,
      created_at: String,
      updated_at: String
  )
  private implicit val todoPublicEncoder: Encoder[TodoPublic] = deriveEncoder
  private def toPublic(t: TodoRec): TodoPublic =
    TodoPublic(t.id, t.title, t.description, t.completed, t.created_at, t.updated_at)

  private case class RegisterReq(username: String, password: String)
  private implicit val registerReqDecoder: Decoder[RegisterReq] = deriveDecoder
  private case class LoginReq(username: String, password: String)
  private implicit val loginReqDecoder: Decoder[LoginReq] = deriveDecoder
  private case class PasswordReq(old_password: String, new_password: String)
  private implicit val passwordReqDecoder: Decoder[PasswordReq] = deriveDecoder
  private case class CreateTodoReq(title: String, description: Option[String])
  private implicit val createTodoReqDecoder: Decoder[CreateTodoReq] = deriveDecoder
  private case class UpdateTodoReq(title: Option[String], description: Option[String], completed: Option[Boolean])
  private implicit val updateTodoReqDecoder: Decoder[UpdateTodoReq] = deriveDecoder

  case class State(
      nextUserId: Int,
      nextTodoId: Int,
      users: Map[Int, UserRec],
      usernameIndex: Map[String, Int],
      sessions: Map[String, Int], // token -> userId
      todos: Map[Int, TodoRec]
  )
  object State {
    def empty: State = State(1, 1, Map.empty, Map.empty, Map.empty, Map.empty)
  }

  private def nowInstant(): java.time.Instant = java.time.Instant.now().truncatedTo(java.time.temporal.ChronoUnit.SECONDS)
  private def nowTs(): String = nowInstant().toString
  private def bumpedUpdateTs(prev: String): String = {
    val now = nowInstant()
    val prevI = scala.util.Try(java.time.Instant.parse(prev)).getOrElse(java.time.Instant.EPOCH)
    val chosen = if (now.isAfter(prevI)) now else prevI.plusSeconds(1)
    chosen.toString
  }

  private val JsonCT: Header.Raw = Header.Raw(CIString("Content-Type"), "application/json")

  private def jsonError(msg: String, status: Status): Response[IO] =
    Response[IO](status).withEntity(Json.obj("error" -> Json.fromString(msg))).putHeaders(JsonCT)

  private def jsonOk(entity: Json, status: Status = Status.Ok): Response[IO] =
    Response[IO](status).withEntity(entity).putHeaders(JsonCT)

  private def readJson[A: Decoder](req: Request[IO]): IO[Either[Response[IO], A]] =
    req.as[Json].attempt.flatMap {
      case Left(_) => IO.pure(Left(jsonError("Invalid JSON", Status.BadRequest)))
      case Right(json) =>
        IO.pure(json.as[A].leftMap(_ => jsonError("Invalid JSON", Status.BadRequest)))
    }

  private def withAuth(state: Ref[IO, State])(req: Request[IO]): IO[Either[Response[IO], (UserRec, String)]] = {
    val tokenOpt = req.cookies.find(_.name == "session_id").map(_.content)
    tokenOpt match {
      case None => IO.pure(Left(jsonError("Authentication required", Status.Unauthorized)))
      case Some(token) =>
        state.get.map { s =>
          s.sessions.get(token).flatMap(uid => s.users.get(uid)) match {
            case Some(user) => Right((user, token))
            case None => Left(jsonError("Authentication required", Status.Unauthorized))
          }
        }
    }
  }

  private def routes(state: Ref[IO, State]): HttpRoutes[IO] = HttpRoutes.of[IO] {

    // POST /register
    case req @ POST -> Root / "register" =>
      readJson[RegisterReq](req).flatMap {
        case Left(err) => IO.pure(err)
        case Right(body) =>
          val usernameValid = body.username.matches("^[a-zA-Z0-9_]{3,50}$")
          val passwordValid = body.password.length >= 8
          if (!usernameValid) IO.pure(jsonError("Invalid username", Status.BadRequest))
          else if (!passwordValid) IO.pure(jsonError("Password too short", Status.BadRequest))
          else state.modify { s =>
            if (s.usernameIndex.contains(body.username))
              (s, Left(jsonError("Username already exists", Status.Conflict)))
            else {
              val id = s.nextUserId
              val user = UserRec(id, body.username, body.password)
              val s2 = s.copy(
                nextUserId = id + 1,
                users = s.users + (id -> user),
                usernameIndex = s.usernameIndex + (body.username -> id)
              )
              (s2, Right(user))
            }
          }.map {
            case Left(err) => err
            case Right(user) =>
              jsonOk(UserPublic(user.id, user.username).asJson, Status.Created)
          }
      }

    // POST /login
    case req @ POST -> Root / "login" =>
      readJson[LoginReq](req).flatMap {
        case Left(err) => IO.pure(err)
        case Right(body) =>
          state.modify { s =>
            s.usernameIndex.get(body.username).flatMap(s.users.get) match {
              case Some(user) if user.password == body.password =>
                val token = UUID.randomUUID().toString.replaceAll("-", "")
                val s2 = s.copy(sessions = s.sessions + (token -> user.id))
                (s2, Right((user, token)))
              case _ => (s, Left(jsonError("Invalid credentials", Status.Unauthorized)))
            }
          }.map {
            case Left(err) => err
            case Right((user, token)) =>
              val cookie = ResponseCookie(name = "session_id", content = token, path = Some("/"), httpOnly = true)
              jsonOk(UserPublic(user.id, user.username).asJson, Status.Ok).addCookie(cookie)
          }
      }

    // POST /logout
    case req @ POST -> Root / "logout" =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((_, token)) =>
          state.update(s => s.copy(sessions = s.sessions - token)) *>
            IO.pure(jsonOk(Json.obj(), Status.Ok))
      }

    // GET /me
    case req @ GET -> Root / "me" =>
      withAuth(state)(req).map {
        case Left(err) => err
        case Right((user, _)) => jsonOk(UserPublic(user.id, user.username).asJson, Status.Ok)
      }

    // PUT /password
    case req @ PUT -> Root / "password" =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          readJson[PasswordReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(body) =>
              if (body.old_password != user.password) IO.pure(jsonError("Invalid credentials", Status.Unauthorized))
              else if (body.new_password.length < 8) IO.pure(jsonError("Password too short", Status.BadRequest))
              else state.update { s =>
                s.copy(users = s.users.updated(user.id, user.copy(password = body.new_password)))
              } *> IO.pure(jsonOk(Json.obj(), Status.Ok))
          }
      }

    // GET /todos
    case req @ GET -> Root / "todos" =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          state.get.map { s =>
            val list = s.todos.values.filter(_.ownerId == user.id).toList.sortBy(_.id).map(toPublic)
            jsonOk(list.asJson, Status.Ok)
          }
      }

    // POST /todos
    case req @ POST -> Root / "todos" =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          readJson[CreateTodoReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(body) =>
              val title = Option(body.title).getOrElse("")
              if (title.trim.isEmpty) IO.pure(jsonError("Title is required", Status.BadRequest))
              else state.modify { s =>
                val id = s.nextTodoId
                val ts = nowTs()
                val todo = TodoRec(id, user.id, title, body.description.getOrElse(""), completed = false, created_at = ts, updated_at = ts)
                val s2 = s.copy(nextTodoId = id + 1, todos = s.todos + (id -> todo))
                (s2, todo)
              }.map { todo => jsonOk(toPublic(todo).asJson, Status.Created) }
          }
      }

    // GET /todos/:id
    case req @ GET -> Root / "todos" / IntVar(id) =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          state.get.map { s =>
            s.todos.get(id) match {
              case Some(t) if t.ownerId == user.id => jsonOk(toPublic(t).asJson, Status.Ok)
              case _ => jsonError("Todo not found", Status.NotFound)
            }
          }
      }

    // PUT /todos/:id
    case req @ PUT -> Root / "todos" / IntVar(id) =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          readJson[UpdateTodoReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(body) =>
              state.modify { s =>
                s.todos.get(id) match {
                  case Some(t) if t.ownerId == user.id =>
                    body.title match {
                      case Some(tl) if tl.trim.isEmpty => (s, Left(jsonError("Title is required", Status.BadRequest)))
                      case _ =>
                        val newTitle = body.title.getOrElse(t.title)
                        val newDesc = body.description.getOrElse(t.description)
                        val newComp = body.completed.getOrElse(t.completed)
                        val ts = if (body.title.isDefined || body.description.isDefined || body.completed.isDefined) bumpedUpdateTs(t.updated_at) else t.updated_at
                        val updated = t.copy(title = newTitle, description = newDesc, completed = newComp, updated_at = ts)
                        (s.copy(todos = s.todos.updated(id, updated)), Right(updated))
                    }
                  case _ => (s, Left(jsonError("Todo not found", Status.NotFound)))
                }
              }.map {
                case Left(err) => err
                case Right(todo) => jsonOk(toPublic(todo).asJson, Status.Ok)
              }
          }
      }

    // DELETE /todos/:id
    case req @ DELETE -> Root / "todos" / IntVar(id) =>
      withAuth(state)(req).flatMap {
        case Left(err) => IO.pure(err)
        case Right((user, _)) =>
          state.modify { s =>
            s.todos.get(id) match {
              case Some(t) if t.ownerId == user.id =>
                (s.copy(todos = s.todos - id), Right(()))
              case _ => (s, Left(jsonError("Todo not found", Status.NotFound)))
            }
          }.map {
            case Left(err) => err
            case Right(_) => Response[IO](Status.NoContent)
          }
      }
  }

  override def run(args: List[String]): IO[ExitCode] = {
    val defaultPort = 8080
    val portNum = args.sliding(2, 1).collect { case List("--port", p) => p }.toList.headOption.map(_.toInt).getOrElse(defaultPort)
    val ip4sPort = Port.fromInt(portNum).getOrElse(port"8080")

    for {
      ref <- Ref.of[IO, State](State.empty)
      httpApp = routes(ref).orNotFound
      _ <- org.http4s.ember.server.EmberServerBuilder.default[IO]
        .withHost(ipv4"0.0.0.0") // ensure 0.0.0.0 bind
        .withPort(ip4sPort)
        .withHttpApp(httpApp)
        .build
        .use { _ =>
          IO.println(s"Server started at http://0.0.0.0:$portNum") *> IO.never
        }
    } yield ExitCode.Success
  }
}

File: data_mcp/gpt-5/go/rep3/main.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"golang.org/x/crypto/bcrypt"
)

const (
	cookieName = "session_id"
	isoLayout  = "2006-01-02T15:04:05Z"
)

type User struct {
	ID       int    `json:"id"`
	Username string `json:"username"`
	// internal only
	PasswordHash []byte `json:"-"`
}

type Todo struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Completed   bool   `json:"completed"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
	// internal
	UserID int `json:"-"`
}

type Store struct {
	mu             sync.RWMutex
	nextUserID     int
	nextTodoID     int
	usersByID      map[int]*User
	usersByName    map[string]*User
	sessions       map[string]int // token -> userID
	todosByID      map[int]*Todo
}

func NewStore() *Store {
	return &Store{
		nextUserID:  1,
		nextTodoID:  1,
		usersByID:   make(map[int]*User),
		usersByName: make(map[string]*User),
		sessions:    make(map[string]int),
		todosByID:   make(map[int]*Todo),
	}
}

// Helpers
func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	enc := json.NewEncoder(w)
	_ = enc.Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func generateToken(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

var usernameRE = regexp.MustCompile(`^[a-zA-Z0-9_]{3,50}$`)

func validateUsername(u string) bool {
	if len(u) < 3 || len(u) > 50 {
		return false
	}
	return usernameRE.MatchString(u)
}

func validatePassword(p string) bool {
	return len(p) >= 8
}

// Authentication helpers
func (s *Store) getUserFromRequest(r *http.Request) (*User, string) {
	cookie, err := r.Cookie(cookieName)
	if err != nil || cookie.Value == "" {
		return nil, ""
	}
	// lookup session
	s.mu.RLock()
	uid, ok := s.sessions[cookie.Value]
	s.mu.RUnlock()
	if !ok {
		return nil, cookie.Value
	}
	s.mu.RLock()
	u := s.usersByID[uid]
	s.mu.RUnlock()
	return u, cookie.Value
}

func (s *Store) requireAuth(w http.ResponseWriter, r *http.Request) (*User, string, bool) {
	u, token := s.getUserFromRequest(r)
	if u == nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return nil, token, false
	}
	return u, token, true
}

// Handlers
func (s *Store) handleRegister(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	var req struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	dec := json.NewDecoder(r.Body)
	if err := dec.Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if !validateUsername(req.Username) {
		writeError(w, http.StatusBadRequest, "Invalid username")
		return
	}
	if !validatePassword(req.Password) {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal server error")
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, exists := s.usersByName[req.Username]; exists {
		writeError(w, http.StatusConflict, "Username already exists")
		return
	}
	id := s.nextUserID
	s.nextUserID++
	u := &User{ID: id, Username: req.Username, PasswordHash: hash}
	s.usersByID[id] = u
	s.usersByName[req.Username] = u
	writeJSON(w, http.StatusCreated, map[string]interface{}{"id": u.ID, "username": u.Username})
}

func (s *Store) handleLogin(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	var req struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	dec := json.NewDecoder(r.Body)
	if err := dec.Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	s.mu.RLock()
	u := s.usersByName[req.Username]
	s.mu.RUnlock()
	if u == nil || bcrypt.CompareHashAndPassword(u.PasswordHash, []byte(req.Password)) != nil {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	tok, err := generateToken(32)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal server error")
		return
	}
	s.mu.Lock()
	s.sessions[tok] = u.ID
	s.mu.Unlock()
	cookie := &http.Cookie{Name: cookieName, Value: tok, Path: "/", HttpOnly: true}
	http.SetCookie(w, cookie)
	writeJSON(w, http.StatusOK, map[string]interface{}{"id": u.ID, "username": u.Username})
}

func (s *Store) handleLogout(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	u, token, ok := s.requireAuth(w, r)
	_ = u
	if !ok {
		return
	}
	if token != "" {
		s.mu.Lock()
		delete(s.sessions, token)
		s.mu.Unlock()
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{})
}

func (s *Store) handleMe(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{"id": u.ID, "username": u.Username})
}

func (s *Store) handlePassword(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPut {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	var req struct {
		OldPassword string `json:"old_password"`
		NewPassword string `json:"new_password"`
	}
	dec := json.NewDecoder(r.Body)
	if err := dec.Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if bcrypt.CompareHashAndPassword(u.PasswordHash, []byte(req.OldPassword)) != nil {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	if !validatePassword(req.NewPassword) {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(req.NewPassword), bcrypt.DefaultCost)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal server error")
		return
	}
	s.mu.Lock()
	u.PasswordHash = hash
	s.mu.Unlock()
	writeJSON(w, http.StatusOK, map[string]interface{}{})
}

func (s *Store) handleTodos(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		s.listTodos(w, r)
	case http.MethodPost:
		s.createTodo(w, r)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func nowISO() string {
	return time.Now().UTC().Format(isoLayout)
}

func (s *Store) listTodos(w http.ResponseWriter, r *http.Request) {
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	// Gather todos for user
	s.mu.RLock()
	list := make([]*Todo, 0)
	for _, t := range s.todosByID {
		if t.UserID == u.ID {
			// copy struct to avoid exposing internal pointer that may be modified concurrently
			copy := *t
			list = append(list, &copy)
		}
	}
	s.mu.RUnlock()
	sort.Slice(list, func(i, j int) bool { return list[i].ID < list[j].ID })
	writeJSON(w, http.StatusOK, list)
}

func (s *Store) createTodo(w http.ResponseWriter, r *http.Request) {
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	var req struct {
		Title       string `json:"title"`
		Description string `json:"description"`
	}
	dec := json.NewDecoder(r.Body)
	if err := dec.Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if strings.TrimSpace(req.Title) == "" {
		writeError(w, http.StatusBadRequest, "Title is required")
		return
	}
	if req.Description == "" {
		// default to empty string already
	}
	created := nowISO()
	s.mu.Lock()
	id := s.nextTodoID
	s.nextTodoID++
	t := &Todo{
		ID:          id,
		Title:       req.Title,
		Description: req.Description,
		Completed:   false,
		CreatedAt:   created,
		UpdatedAt:   created,
		UserID:      u.ID,
	}
	s.todosByID[id] = t
	s.mu.Unlock()
	writeJSON(w, http.StatusCreated, t)
}

func parseTodoID(path string) (int, error) {
	// Expect /todos/:id exactly, no trailing segments
	parts := strings.Split(strings.Trim(path, "/"), "/")
	if len(parts) != 2 || parts[0] != "todos" {
		return 0, errors.New("not a todo item path")
	}
	id, err := strconv.Atoi(parts[1])
	if err != nil {
		return 0, err
	}
	return id, nil
}

func (s *Store) handleTodoItem(w http.ResponseWriter, r *http.Request) {
	id, err := parseTodoID(r.URL.Path)
	if err != nil {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	switch r.Method {
	case http.MethodGet:
		s.getTodo(w, r, id)
	case http.MethodPut:
		s.updateTodo(w, r, id)
	case http.MethodDelete:
		s.deleteTodo(w, r, id)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func (s *Store) findUserTodo(u *User, id int) (*Todo, bool) {
	s.mu.RLock()
	t, ok := s.todosByID[id]
	if ok && t.UserID != u.ID {
		ok = false
		t = nil
	}
	s.mu.RUnlock()
	return t, ok
}

func (s *Store) getTodo(w http.ResponseWriter, r *http.Request, id int) {
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	t, ok := s.findUserTodo(u, id)
	if !ok {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	// return a copy
	copy := *t
	writeJSON(w, http.StatusOK, &copy)
}

func (s *Store) updateTodo(w http.ResponseWriter, r *http.Request, id int) {
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	// check existence/ownership
	s.mu.RLock()
	t, exists := s.todosByID[id]
	ownerOK := exists && t.UserID == u.ID
	s.mu.RUnlock()
	if !ownerOK {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	// decode into map to detect presence
	var raw map[string]json.RawMessage
	dec := json.NewDecoder(r.Body)
	if err := dec.Decode(&raw); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	var newTitle *string
	var newDesc *string
	var newCompleted *bool
	if v, ok := raw["title"]; ok {
		var sVal string
		if err := json.Unmarshal(v, &sVal); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		if strings.TrimSpace(sVal) == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		newTitle = &sVal
	}
	if v, ok := raw["description"]; ok {
		var sVal string
		if err := json.Unmarshal(v, &sVal); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		newDesc = &sVal
	}
	if v, ok := raw["completed"]; ok {
		var bVal bool
		if err := json.Unmarshal(v, &bVal); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		newCompleted = &bVal
	}
	// apply updates
	s.mu.Lock()
	if newTitle != nil {
		t.Title = *newTitle
	}
	if newDesc != nil {
		t.Description = *newDesc
	}
	if newCompleted != nil {
		t.Completed = *newCompleted
	}
	t.UpdatedAt = nowISO()
	s.mu.Unlock()
	// return copy
	copy := *t
	writeJSON(w, http.StatusOK, &copy)
}

func (s *Store) deleteTodo(w http.ResponseWriter, r *http.Request, id int) {
	u, _, ok := s.requireAuth(w, r)
	if !ok {
		return
	}
	s.mu.Lock()
	t, exists := s.todosByID[id]
	if !exists || t.UserID != u.ID {
		s.mu.Unlock()
		// Return 404 with JSON error
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	delete(s.todosByID, id)
	s.mu.Unlock()
	// DELETE success: 204 no body
	w.WriteHeader(http.StatusNoContent)
}

// Custom mux
type apiServer struct {
	store *Store
	mux   *http.ServeMux
}

func newAPIServer(store *Store) *apiServer {
	mux := http.NewServeMux()
	s := &apiServer{store: store, mux: mux}
	mux.HandleFunc("/register", store.handleRegister)
	mux.HandleFunc("/login", store.handleLogin)
	mux.HandleFunc("/logout", store.handleLogout)
	mux.HandleFunc("/me", store.handleMe)
	mux.HandleFunc("/password", store.handlePassword)
	// Route /todos and /todos/:id using separate patterns
	mux.HandleFunc("/todos", store.handleTodos)
	mux.HandleFunc("/todos/", store.handleTodoItem)
	// NotFound handler override
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		writeError(w, http.StatusNotFound, "Not found")
	})
	return s
}

func (s *apiServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// Ensure content-type for all JSON responses. We'll set in handlers when writing.
	s.mux.ServeHTTP(w, r)
}

func main() {
	port := flag.Int("port", 8080, "port to listen on")
	flag.Parse()
	addr := fmt.Sprintf("0.0.0.0:%d", *port)
	store := NewStore()
	server := newAPIServer(store)
	// Use a server with reasonable timeouts
	hs := &http.Server{
		Addr:              addr,
		Handler:           server,
		ReadHeaderTimeout: 5 * time.Second,
		ReadTimeout:       10 * time.Second,
		WriteTimeout:      10 * time.Second,
		IdleTimeout:       60 * time.Second,
	}
	log.Printf("Server listening on %s", addr)
	if err := hs.ListenAndServe(); err != nil && err != http.ErrServerClosed {
		log.Fatalf("server error: %v", err)
	}
}
